parameter/criterion/src/Criterion.cpp

```diff
@@ -0,0 +1,296 @@
+    }
+    // Track the number of modifications for this criterion
+    _uiNbModifications++;
+}
+
+CriterionInterface::State Criterion::getState() const
+{
+    return mState;
+}
+
+std::string Criterion::getCriterionName() const
+{
+    return mName;
+}
+
+std::string Criterion::getFormattedDescription(bool bWithTypeInfo, bool bHumanReadable) const
+{
+    std::string strFormattedDescription;
+    std::string typeName = isInclusive() ? "Inclusive" : "Exclusive";
+
+    if (bHumanReadable) {
+
+        if (bWithTypeInfo) {
+
+            // Display type info
+            CUtility::appendTitle(strFormattedDescription, mName + ":");
+
+            // States
+            strFormattedDescription += "Possible states ";
+
+            // Type Kind
+            strFormattedDescription += "(";
+            strFormattedDescription += typeName;
+            strFormattedDescription += "): ";
+
+            // States
+            strFormattedDescription += listPossibleValues() + "\n";
+
+            // Current State
+            strFormattedDescription += "Current state";
+        } else {
+            // Name only
+            strFormattedDescription = mName;
+        }
+
+        // Current State
+        strFormattedDescription += " = " + getFormattedState();
+    } else {
+        // Name
+        strFormattedDescription = "Criterion name: " + mName;
+
+        if (bWithTypeInfo) {
+            // Type Kind
+            strFormattedDescription += ", type kind: ";
+            strFormattedDescription +=  typeName;
+        }
+
+        // Current State
+        strFormattedDescription += ", current state: " + getFormattedState();
+
+        if (bWithTypeInfo) {
+            // States
+            strFormattedDescription += ", states: " + listPossibleValues();
+        }
+    }
+    return strFormattedDescription;
+}
+
+void Criterion::toXml(CXmlElement& xmlElement, CXmlSerializingContext& serializingContext) const
+{
+    (void)serializingContext;
+    xmlElement.setAttributeString("Value", getFormattedState());
+    xmlElement.setAttributeString("Name", mName);
+    xmlElement.setAttributeString("Kind", isInclusive() ? "Inclusive" : "Exclusive");
+
+    for (auto& valuePair : mValues) {
+        CXmlElement childValuePairElement;
+
+        xmlElement.createChild(childValuePairElement, "ValuePair");
+        childValuePairElement.setAttributeString("Literal", valuePair.first);
+        childValuePairElement.setAttributeSignedInteger("Numerical", valuePair.second);
+    }
+}
+
+bool Criterion::isInclusive() const
+{
+    return false;
+}
+
+bool Criterion::getNumericalValue(const std::string& literalValue,
+                                  int& numericalValue) const
+{
+    try {
+        numericalValue = mValues.at(literalValue);
+        return true;
+    }
+    catch (std::out_of_range& e) {
+        return false;
+    }
+}
+
+bool Criterion::getLiteralValue(int numericalValue, std::string& literalValue) const
+{
+    for (auto& value : mValues) {
```

Comment:
find only works with "key" map, as we search for the mapped value I think we need to test all keys